### ISSUE TITLE
Rule proposal: compat/special-param-shadow

### ISSUE THREAD

**Issue by ss-o** [z-shell/zsh-lint#64]
## Rule ID

`compat/special-param-shadow`

## Rule name

Shadowing special shell parameters

## Summary

Reports `local`/`typeset` declarations (and assignments in function scope) that shadow parameters set by the shell — e.g. `ZSH_VERSION`, `OSTYPE`, `pipestatus` — changing what every function called from that scope observes.

## Why this matters

`zshparam` §Parameters Set By The Shell documents these as shell-provided state. Version probes like `is-at-least $ZSH_VERSION` are pervasive in plugin code; a `local ZSH_VERSION=...` in a caller silently feeds the override to all nested code for the lifetime of the scope. Unlike ordinary shadowing, the reader has no declaration of the original to look up — the shell set it.

## Corpus evidence (2026-06-12, LangZsh-parseable files)

- `zd/docker/utils.zsh:78` — `local ZSH_VERSION="$1"` (a helper that takes a target version as its first argument and stores it in the special parameter's name).

## Bad / Good examples

Bad:

```zsh
compile_zsh() {
  local ZSH_VERSION="$1"   # nested calls now see the fake version
  ...
}
```

Good:

```zsh
compile_zsh() {
  local target_zsh_version="$1"
  ...
}
```

## Proposed severity

`warning`

## Severity justification

Per the mapping, `warning` covers likely-bug patterns with realistic false-positive potential: the corpus instance is functional-but-misleading, and deliberate shims exist.

## False-positive analysis

1. **Deliberate compatibility shims / test harnesses** that fake `ZSH_VERSION` or `OSTYPE` for downstream code: that is exactly the rule's target behavior made intentional — suppress with a reason.
2. **Curated list, not a heuristic**: v1 flags only a documented set of read-mostly shell-set parameters (`ZSH_VERSION`, `ZSH_PATCHLEVEL`, `ZSH_NAME`, `ZSH_ARGZERO`, `OSTYPE`, `MACHTYPE`, `VENDOR`, `pipestatus`, `status`). Conventionally-mutated specials (`path`, `fpath`, `PATH`, `REPLY`, `match`) are excluded by design — mutating those is the documented way to use them.
3. **Reads are fine**: only declarations/assignments are flagged, never expansions.

## Parser dependencies

None — the evidence file parses under mvdan/sh LangZsh v3.13.1.

## Suppression

`# zsh-lint disable=compat/special-param-shadow -- <reason>` per `docs/project/suppression.md`.

**Comment by ss-o** [comment]
Implemented and merged into `next` via #76 (merge `cd9f4c1`).

Final v1 semantics follow the official Zsh builtin behavior, including ordered `g`/`m`/`x` options, `GLOBAL_EXPORT`, standalone `+` display mode, separate numeric option arguments, quoted/escaped words, and `typeset -T` operand roles. In particular, active `-m` with effective `+g` creates local matches and is reportable for exact curated names.

All Go CI, generated-doc, and Trunk checks passed; independent review finished with no findings. Deferred bare-assignment and `integer`/`float` work remains tracked in #72.
